### DISCUSSION TITLE
Would you (or do you) use `Option` to model nullable database columns?

### DISCUSSION THREAD

**Discussion by Meigs2** [louthy/language-ext#1102] [Q&A] [answered]
I'm designing some database models for a new project and was wondering if anyone uses `Option` (or similar structure) to model nullable database column types instead of using nullable reference types?

I can see the benefit of having `Option` directly in the model, however as someone still very new to functional programming, I feel like nullable reference types are more fitting on the raw data model, given nullable reference types in C# seem like they would more directly reflect null data on the database-side rather than `Option`. I think there's an easier argument for having `Option` on something like a DTO or command/query object with MediatR or similar.

If you do use `Option` (Or a similar structure) to represent null, how do you integrate these properties with ORM's like EF Core/Dapper/etc?

Thanks!

**Comment by YuriiNskyi** [comment, marked as answer]
> I feel like nullable reference types are more fitting on the raw data model, given nullable reference types in C# seem like they would more directly reflect null data on the database-side rather than Option.

You have answered the question by yourself. EF Core and other ORM`s are intended to work with raw C# features, so there is no point trying to tie EF Core and high-level functional (unknown for him) things.

In other words, let the `Option` be inside your `Domain`, while nullable reference types remain inside `Infrastructure` (ORMs and so forth). Don't mix them (use `Option` inside `Infrastructure` and vice versa), it is a guaranteed way for a lot of mistakes.

So, how to connect your `Domain` and `Infrastructure`, if they use different concepts for the same abstraction? The answer is pretty simple - use mapping layer. Translations between `null` and `Option` are deadly simple using already provided by this library functionality. Additionally, mapping layer has a great advantage of reducing coupling between your layers (because sometimes layers tend to be changing out of sync of each other).

To summarize, my best advice is even beyond using `Option` - do not use anything such specific inside your `Infrastructure` layer. it's not worth it. Instead of fixing tons of compatibility issues, it is always better to avoid them at all. Raw C# is already almost perfect for such kind of tasks.

Hope that I answered your question!

**Reply by Meigs2** [reply]
Great answer. Thank you!

**Comment by alarya** [comment]
@Meigs2 - I think you might also want to check out the discussion here -  [issue](https://github.com/louthy/language-ext/issues/456)
